network/tests/topology.rs

```diff
@@ -0,0 +1,301 @@
+const MIN_PEERS: u16 = 5;
+const MAX_PEERS: u16 = 10;
+
+async fn test_nodes(n: usize, setup: TestSetup) -> Vec<Node> {
+    let mut nodes = Vec::with_capacity(n);
+
+    for _ in 0..n {
+        let environment = test_environment(setup.clone());
+        let mut node = Node::new(environment).await.unwrap();
+
+        node.establish_address().await.unwrap();
+        nodes.push(node);
+    }
+
+    nodes
+}
+
+async fn start_nodes(nodes: &[Node]) {
+    for node in nodes {
+        // Nodes are started with a slight delay to avoid having peering intervals in phase (this
+        // is the hypothetical real-world worst case scenario).
+        tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+        node.start_services().await;
+    }
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn spawn_nodes_in_a_line() {
+    let setup = TestSetup {
+        consensus_setup: None,
+        peer_sync_interval: 1,
+        ..Default::default()
+    };
+    let mut nodes = test_nodes(N, setup).await;
+    connect_nodes(&mut nodes, Topology::Line).await;
+    start_nodes(&nodes).await;
+
+    // First and Last nodes should have 1 connected peer.
+    wait_until!(
+        5,
+        nodes.first().unwrap().peer_book.read().number_of_connected_peers() == 1
+    );
+    wait_until!(
+        5,
+        nodes.last().unwrap().peer_book.read().number_of_connected_peers() == 1
+    );
+
+    // All other nodes should have two.
+    for i in 1..(nodes.len() - 1) {
+        wait_until!(5, nodes[i].peer_book.read().number_of_connected_peers() == 2);
+    }
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn spawn_nodes_in_a_ring() {
+    let setup = TestSetup {
+        consensus_setup: None,
+        peer_sync_interval: 1,
+        ..Default::default()
+    };
+    let mut nodes = test_nodes(N, setup).await;
+    connect_nodes(&mut nodes, Topology::Ring).await;
+    start_nodes(&nodes).await;
+
+    for node in &nodes {
+        wait_until!(5, node.peer_book.read().number_of_connected_peers() == 2);
+    }
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn spawn_nodes_in_a_star() {
+    let setup = TestSetup {
+        consensus_setup: None,
+        peer_sync_interval: 1,
+        ..Default::default()
+    };
+    let mut nodes = test_nodes(N, setup).await;
+    connect_nodes(&mut nodes, Topology::Star).await;
+    start_nodes(&nodes).await;
+
+    let hub = nodes.first().unwrap();
+    wait_until!(10, hub.peer_book.read().number_of_connected_peers() as usize == N - 1);
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn spawn_nodes_in_a_mesh() {
+    let setup = TestSetup {
+        consensus_setup: None,
+        peer_sync_interval: 5,
+        min_peers: MIN_PEERS,
+        max_peers: MAX_PEERS,
+        ..Default::default()
+    };
+    let mut nodes = test_nodes(N, setup).await;
+    connect_nodes(&mut nodes, Topology::Mesh).await;
+    start_nodes(&nodes).await;
+
+    // Set the sleep interval to 200ms to avoid lock issues.
```

Comment:
```suggestion
    // Set the sleep interval to 200ms to avoid locking issues.
```
We should also add (can be in a follow-up) tests that would check the worst-case peering conditions so that we're sure that those issues result only in a single update slowdown / skip as opposed to anything more problematic.